Feature/eeprom

wvarty opened the pull request.

wvarty commented:
This will need a thorough test before merging.

wvarty commented:
One thing I forgot to add to the eeprom struct was a CRC... I havent used one in past projects, and it's been fine, but I noticed that the chorus32 eeprom struct uses one, and I think when I last spoke to @AlessandroAU and @JyeSmith about this, they were pretty firm on needing it?

JyeSmith commented:
@wvarty previously I have only used the version number as a check.  Worse case (or maybe best) is that it is corrupt and the UIDs are different so the quad wont bind.

![image](https://user-images.githubusercontent.com/14170229/90301411-09b4ae80-dee3-11ea-8bc7-66bceda74a71.png)

wvarty commented:
@JyeSmith I guess worst case is prob more like we read a corrupt value for power output, and someone goes to fly freestyle thinking they are on 200mw, but are actually on 25mw, and totals a quad...

JBKingdon commented:
How often do we expect the values to be written, do we need to worry about wearing out the flash?

wvarty commented:
Everytime we move the scroll wheel in LUA, it will do a write... I dislike how we send the value change to the TX from the handset on scroll... IMO it should only get sent when you press the return button

wvarty commented:
The only way I can think of to reduce write cycles is to put some hysteresis around detecting changes... I.e in the main loop you have a func that runs every 10s to see if the current vals != saved vals, and do the write. Prob with that is, you might have already armed by that time, and we need to stop the timer to do a write.

maybenikhil commented:
@wvarty what if there was a button that saved (could be put in place of the bind button for now)

JBKingdon commented:
Looks awesome. I like Nikhil's idea for a manual save UI element ("save as defaults"), but we also talked about auto save on exit from the lua script which would be more convenient in a lot of cases.

JyeSmith commented:
> 
> 
> Everytime we move the scroll wheel in LUA, it will do a write... I dislike how we send the value change to the TX from the handset on scroll... IMO it should only get sent when you press the return button

This gets my vote.

JyeSmith commented on src/src/config.h:
```diff
@@ -0,0 +1,40 @@
+#include "elrs_eeprom.h"
+
+#define CONFIG_VERSION          1
+#define RESERVED_EEPROM_SIZE    512
```
Already defined in elrs_eeprom.h.  Should it still be here?

wvarty replied:
No, it should not.

wvarty commented:
Horribly broken on STM :( dont merge

JBKingdon commented on src/lib/EEPROM/elrs_eeprom.cpp:
```diff
@@ -18,21 +19,32 @@ ELRS_EEPROM::Begin()
     EEPROM.write(address, value);
+
+#ifdef PLATFORM_ESP32
+    if (!EEPROM.commit())
```
We probably don't want per-byte commit. Can we move the commits up to a transactional level so that there's just one per save?

JBKingdon replied:
sweet

wvarty commented:
Ready for another round of review. Changes were:
- Changed templates to void, and removed the ref symbol
- Added eeprom address to STM begin()
- Split out the Commit() call as per James's suggestion, and only call after all bytes have been written.
- Increased the reserved eeprom size from 512 to 2048 **NOTE: I HAVE NO IDEA WHAT THIS SHOULD ACTUALLY BE??**

AlessandroAU commented on src/src/config.cpp:
```diff
@@ -0,0 +1,112 @@
+}
+
+// Setters
+void ICACHE_RAM_ATTR
```
we don't actually need ICACHE_RAM_ATTR unless we are calling from ISR

JBKingdon commented:
As discussed, just a reminder that it would be great if we could use the put & get templates from the esp32 impl as then we would have a consistent API on the two platforms plus benefit from the 'don't write if the data hasn't changed' behaviour.

JyeSmith commented on src/lib/EEPROM/elrs_eeprom.cpp:
```diff
@@ -1,38 +1,51 @@
 
 #ifdef PLATFORM_STM32
-extEEPROM EEPROM(kbits_2, 1, 1);
+    extEEPROM EEPROM(kbits_2, 1, 1, 0x51);
```
IIRC the r9mm uses the default address 0x50.  Something to keep in mind when reusing this lib for the receivers.

JyeSmith commented on src/lib/EEPROM/elrs_eeprom.cpp:
```diff
@@ -1,38 +1,51 @@
 #ifdef PLATFORM_STM32
-    Wire.setSDA(GPIO_PIN_SDA); // set is needed or it wont work :/
-    Wire.setSCL(GPIO_PIN_SCK);
-    Wire.begin();
```
We may need to keep set and begin.  The r9 receivers do not use the DAC lib where Wire is initialised.  I dont think there is any problem with calling begin twice, but needs to be checked.

JyeSmith commented:
Bench tested with r9 and esp 2.4 gear.  No probs.

JBKingdon commented on src/src/config.h:
```diff
@@ -0,0 +1,40 @@
+    uint32_t    rate;
+    uint32_t    tlm;
+    uint32_t    power;
+    bool        modified;
```
uh, do we want the modified flag to be written to the flash? This feels like it should be just a runtime thing that would be a seperate bool in the class

wvarty replied:
Good point. This should be split out.

JBKingdon commented:
Looks good. Maybe a bit short of doc/comments but the code is pretty readable.

wvarty commented:
> Looks good. Maybe a bit short of doc/comments but the code is pretty readable.

Hmm, I thought it was fairly comment rich... I have no problems adding more if you think it will be useful to others reading this code later?

JBKingdon commented:
> Hmm, I thought it was fairly comment rich... I have no problems adding more if you think it will be useful to others reading this code later?

So, not sure what I was looking at now that made me say that. A quick look again shows it's fine. Sorry.

JyeSmith commented on src/src/config.h:
```diff
@@ -0,0 +1,39 @@
+
+#define CONFIG_VERSION          1
+
+typedef struct {
```
Because a struct is being written and we don't need to track addresses, the variable types can be changed easily in the struct.  Is it worth while dropping them down to uint8_t and reduce the commit time?

wvarty replied:
Replying here for completeness: I don't think its worth a millisecond or two in reducing these types, and it will likely cause pain later when we wanna bump up the size again, and would have to wipe people's settings as part of doing so.

JyeSmith commented:
Tested r9m to r9mm and esp 2.4 gear.  No probs with this PR.

Nice work :fireworks: :fireworks: :fireworks:

raffxj commented:
guys

not working on mine. 'FW update error.no sync'

thanks 👍
